File: 2022/july/11-07-2022/binary-tree-right-side-view.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        ans = []

        def dfs(node=root, level=1):
            if not node: return

            if len(ans) < level:
                ans.append(node.val)
            dfs(node.right, level + 1)
            dfs(node.left, level + 1)

            return

        dfs()
        return ans

File: 2022/july/11-07-2022/binary-tree-right-side-view.rs
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn right_side_view(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut ans = vec![];
        let mut bfs = vec![];
        match root {
            Some(x) => bfs.push(x),
            None => ()
        }
        while bfs.len() != 0 {
            ans.push(bfs[bfs.len()-1].borrow().val);
            let mut row = vec![];
            for node in bfs.iter() {
                match node.borrow().left.as_ref() {
                    // only push onto row if the node actually contains something.
                    Some(x) => row.push(Rc::clone(x)),
                    None => (),
                };
                match node.borrow().right.as_ref() {
                    Some(x) => row.push(Rc::clone(x)),
                    None => (),
                };
            };
            bfs = row;
        }
        ans
    }
}
